### PR TITLE
SyncUploadState

### LINKED ISSUE #47: Need API for Status Access at activity level for UploadRequest
We will need API to have access to the status from UploadRequest at the activity level, so we can handle UI block till Upload is processed successfully and also handle the failure case.

@MJ1998

### PR DESCRIPTION
**Pull request by MJ1998**
Fixes #47 

`SensingEngine.syncUpload` is replaced by `SensingSynchronizer.synchornize` API

in synchronize we
1. Fetch all upload requests 
2. invoke upload on them
3. Collect individual `UploadResult`s  and creat`SyncUploadState`s 
We start off with emitting `SyncUploadState.Started`.
Then the states are `SyncUploadState.InProgress` for all individual `UploadResult`s statuses except for Failed state. Here we emit `SyncUploadState.Failed` and the worker cancels the upload job.
Finally we emit `SyncUploadState.Completed`

We also added `SyncUploadState.NoOp` for the case when a new worker attempts to sync when there is already one worker which is running.

These `SyncUploadState`es are collected by worker and persisted as WorkManager's progress data using `setProgress` API. This progress can be fetched as a flow using 
```
WorkManager.getInstance(context)
      .getWorkInfosForUniqueWorkLiveData(uniqueWorkerName)
      .asFlow()
```

TESTED with code in demo app: Sensory. Look at the following progress bar.
Screenshot 2023-12-09 at 14 48 55

### CODE REVIEW

File: Sensory/src/main/java/com/google/android/sensory/HomeFragment.kt [hunk 1]
```diff
@@ -81,6 +85,59 @@ class HomeFragment : Fragment() {
...
+        }
+      }
+    }
+  }
+
+  private fun showSyncBanner(syncUploadState: SyncUploadState) {
+    with(binding.uploadLayout) {
+      if (linearLayoutUploadStatus.visibility != View.VISIBLE) {
+        // may add fade in animation here later
+        linearLayoutUploadStatus.visibility = View.VISIBLE
```
Comment (chandrashekar-s) [inline Sensory/src/main/java/com/google/android/sensory/HomeFragment.kt:108]: I am assuming this is the case when we start the upload, in that case do we just show the animation? How about the percentage completion?

File: Sensory/src/main/java/com/google/android/sensory/HomeFragment.kt [hunk 2]
```diff
@@ -81,6 +85,59 @@ class HomeFragment : Fragment() {
...
+
+  private fun hideSyncBanner(syncUploadState: SyncUploadState) {
+    if (syncUploadState is SyncUploadState.Completed) {
+      updateUploadPercent(syncUploadState.totalRequests, syncUploadState.totalRequests)
+      binding.uploadLayout.uploadPercent.text = "Uploaded"
+      binding.uploadLayout.linearLayoutUploadStatus.visibility = View.GONE
+    } else if (syncUploadState is SyncUploadState.Failed) {
+      binding.uploadLayout.uploadPercent.text = "Failed"
+      // we can add some animation here
+      binding.uploadLayout.linearLayoutUploadStatus.visibility = View.GONE
```
Comment (chandrashekar-s) [inline Sensory/src/main/java/com/google/android/sensory/HomeFragment.kt:123]: How about providing an error message as well?

**Review by chandrashekar-s — COMMENTED** [review]
Thanks for the changes @MJ1998, they look fine. Only a couple of minor comments. Please have a look

File: Sensory/src/main/java/com/google/android/sensory/HomeFragment.kt (hunk 2 above)
Comment (MJ1998) [reply Sensory/src/main/java/com/google/android/sensory/HomeFragment.kt:125]: We give a toast saying "Upload Failed"

File: Sensory/src/main/java/com/google/android/sensory/HomeFragment.kt (hunk 1 above)
Comment (MJ1998) [reply Sensory/src/main/java/com/google/android/sensory/HomeFragment.kt:109]: Yeah there will be 0% progress shown.

File: sensing/src/main/java/com/google/android/sensing/upload/SensingSynchronizer.kt [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+   * [issue](https://github.com/google/android-fhir/issues/2119).
+   */
+  suspend fun synchronize(): Flow<SyncUploadState> = flow {
+    val uploadRequestList =
+      (sensingEngine.listUploadRequest(status = RequestStatus.UPLOADING) +
+        sensingEngine.listUploadRequest(status = RequestStatus.PENDING))
+
+    var lastSyncUploadState = SyncUploadState.Started(uploadRequestList.size) as SyncUploadState
+    emit(lastSyncUploadState)
+    if (uploadRequestList.isNotEmpty()) {
```
Comment (chandrashekar-s) [inline sensing/src/main/java/com/google/android/sensing/upload/SensingSynchronizer.kt:49]: What if the `uploadRequestList.isNotEmpty()` is empty? Should we just complete it immediately?

**Review by chandrashekar-s — APPROVED** [review]
The changes look fine @MJ1998. There is only minor comment. You can address that and can merge the changes.

File: sensing/src/main/java/com/google/android/sensing/upload/SensingSynchronizer.kt (hunk 3 above)
Comment (MJ1998) [reply sensing/src/main/java/com/google/android/sensing/upload/SensingSynchronizer.kt:49]: Done!

Added lines
```
else { // no requests to process
      emit(SyncUploadState.Completed(0))
    }
```

**Pull request merged**